rfcs/records.md

```diff
@@ -0,0 +1,448 @@
+keep the method calls as efficient as possible, as they don't need to check whether the record has a given method as a field.
+
+> TODO: Should we use `__namecall` instead of raw MT access? It seems more consistent, but at the same time `__namecall` today expects a function so it might be best
+> to leave it as is?
+
+> TODO: How do we expose the record keys? Should it be a builtin? Accessible through shape? Not available initially?
+
+### Defining and constructing records
+
+To define a record, you need to create the shape, which you can do using the newly introduced syntax with a context-sensitive `record` keyword.
+
+> TODO: The draft RFC suggests two options for the syntax; only one will be chosen in the final version
+
+Syntax A:
+
+```lua
+record Person = { name: string, age: number }
+
+-- types can be omitted and default to any
+record Point = { x, y }
+```
+
+Syntax B:
+
+```lua
+record Person(name: string, age: number)
+-- types can be omitted and default to any
+record Point(x, y)
+```
+
+This defines `Point` simultaneously as a local variable that corresponds to the shape table, and a type variable that corresponds to the record type.
+
+The resulting shape table automatically is set up to be a valid record shape, but can still be modified by adding methods to it:
+
+```lua
+function Point.__add(l, r)
+    return Point(l.x + r.x, l.y + r.y)
+end
+
+function Point:sum()
+    return self.x + self.y
+end
+
+function Point.newDiagonal(v)
+    return Point(v, v)
+end
+```
+
+Note that `Point` is simply a table and as such it can be used to store static methods as well; as it also serves as a metatable, metafields defined on this table
+will change the behavior of the record values. The shape isn't frozen automatically but can be frozen manually if desired via `table.freeze`.
+
+To create a record, you need to use a record constructor. This is done using call-like syntax:
+
+Syntax A:
+
+```lua
+local person = Person { name = "Bob", age = 42 }
+```
+
+Syntax B:
+
+```lua
+local person = Person("Bob", 42)
+```
+
+The big difference between variants A and B is whether you need to spell out field names at construction time. There's precedents for going either way; some
+languages like F#/C#/Kotlin implement record construction as a function call and when you define a record, you essentially define the record constructor. This
+is beautifully concise, but is a bit more difficult to migrate away from tables, and it's easy to mix up the names. Variant A is more verbose and requires a bit
+more magic at compile time to keep construction efficient.
+
+In variant A, it would make sense to allow omission of any field, in which case it gets replaced with the default of `nil`. A future extension (not part of this RFC) could be made to
+allow specification of default values at record definition time. Type checker would fail to type check record construction if fields that have non-optional types
+have the values omitted.
```

Comment:
To create a record with all fields omitted, would I write `Person {}`?

Reply:
Yeah.